### PR TITLE
[No QA] Asking for company name and URL during invoice creation

### PR DESCRIPTION
**Pull request by VickyStash**
### Details

Updates the invoice creation flow to support adding company name and URL if it wasn't added before.

### Fixed Issues

$ https://github.com/Expensify/App/issues/45174
$ https://github.com/Expensify/App/issues/45173
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

**!Pre-step:** Update [hasInvoicingDetails check](https://github.com/Expensify/App/pull/46009/files#diff-c3278b76b0699dc5c04d47722c67669ef432ee715141b506716a6be212eb3f08R208-R212) to always return false so you can always see the expected screen. Please, remember that API doesn't support adding company name and the website for now.

1. Press FAB -> Send Invoice
2. Go through the steps to reach the confirmation screen. Make sure you see the `Next` button instead of Send Invoice.
3. Tap on the `Next`. You should appear on the Company info page. Make sure it looks as expected.
4. Make sure the validation on the page works as expected.
5. Tap `Send invoice` -> the invoice should be successfully created.

### Offline tests

Same as in the **Tests** section

### QA Steps

- [x] Verify that no errors appear in the JS console

N/A

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C01GTK53T8Q/p1721825016527359
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/52f7d703-e7af-414f-a562-4d1591daf1b8

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/4d0bf04a-386c-4574-9a71-4c3e9c65e471

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/94adf916-0f4e-4c30-b9ad-df272d09bdeb

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/34835217-9f25-4e07-9df4-1f4b45de2f08

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/db013039-f227-4f87-8f5e-3bfdfb29073d

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/35e91340-cfcd-4ef4-b349-e81295da6646

</details>

### CODE REVIEW

File: src/languages/es.ts [hunk 1]
```diff
@@ -705,6 +705,13 @@ export default {
         receiptStatusText: 'Solo tú puedes ver este recibo cuando se está escaneando. Vuelve más tarde o introduce los detalles ahora.',
         receiptScanningFailed: 'El escaneo de recibo ha fallado. Introduce los detalles manualmente.',
         transactionPendingDescription: 'Transacción pendiente. Puede tardar unos días en contabilizarse.',
+        companyInfo: 'Información de la empresa',
+        companyInfoDescription: 'Necesitamos algunos detalles más antes de que pueda enviar su primera factura.',
+        yourCompanyName: 'Nombre de su empresa',
+        yourCompanyWebsite: 'Sitio web de su empresa',
+        yourCompanyWebsiteNote: 'Si no tiene un sitio web, puede proporcionar el perfil de LinkedIn o de las redes sociales de su empresa.',
+        invalidDomainError: 'Ha introducido un dominio no válido. Para continuar, introduzca un dominio válido.',
+        publicDomainError: 'Ha introducido un dominio público. Para continuar, introduzca un dominio privado.',
```
Comment (VickyStash) [inline src/languages/es.ts:714]: I've asked for translations approve in the slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1721825016527359

File: src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    const currentUserPersonalDetails = useCurrentUserPersonalDetails();
+
+    const [policy] = useOnyx(`${ONYXKEYS.COLLECTION.POLICY}${IOU.getIOURequestPolicyID(transaction, report)}`);
+    const [policyCategories] = useOnyx(`${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${IOU.getIOURequestPolicyID(transaction, report)}`);
+    const [policyTags] = useOnyx(`${ONYXKEYS.COLLECTION.POLICY_TAGS}${IOU.getIOURequestPolicyID(transaction, report)}`);
+
+    const formattedAmount = CurrencyUtils.convertToDisplayString(Math.abs(transaction?.amount ?? 0), transaction?.currency);
+
+    const extractUrlDomain = (url: string): string | undefined => {
+        const DOMAIN_BASE_REGEX = '^(?:https?:\\/\\/)?(?:www\\.)?([^\\/]+)';
```
Comment (getusha) [inline src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx:46]: I wonder if we already have similar util function somewhere? if not, i think it's better to move the regex to `CONST.js`  and move the function to `libs/Url.ts`

File: src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        const match = String(url).match(DOMAIN_BASE_REGEX);
+
+        if (!match) {
+            return undefined;
+        }
+
+        return match[1];
```
Comment (getusha) [inline src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx:53]: NAB:
```suggestion
        return match?.[1];
```

File: src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx (hunk 2 above)
Comment (getusha) [reply src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx:46]: Got my answer from here https://github.com/Expensify/App/issues/45174#issuecomment-2248100443 👍

Comment (getusha) [reply src/pages/iou/request/step/IOURequestStepCompanyInfo.tsx:46]: Got my answer from here https://github.com/Expensify/App/issues/45174#issuecomment-2248100443 👍

**Comment by getusha** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by VickyStash** [comment]
@getusha I've applied the requested updates, please take another look

**Comment by VickyStash** [comment]
@getusha kind bump 🙂

**Comment by madmax330** [comment]
@getusha can you prioritize getting this done please?

**Comment by getusha** [comment]
> @getusha can you prioritize getting this done please?

👍 will be on it in 2 hours.

**Comment by getusha** [comment]
@VickyStash seems like we don't allow urls without protocols, and the error message doesn't looks like misleading. should we have a new copy for this? wdyt @VickyStash?

Screenshot 2024-07-30 at 10 15 47 at night

**Comment by VickyStash** [comment]
> @VickyStash seems like we don't allow urls without protocols, and the error message doesn't looks like misleading. should we have a new copy for this? wdyt @VickyStash?
> 
> Screenshot 2024-07-30 at 10 15 47 at night

I've reused the existing [isValidWebsite](https://github.com/Expensify/App/blob/8e79c59615feb7de82e4aaa42d8b2297542f5b3b/src/libs/ValidationUtils.ts#L236-L243) check for the website validity and the existing error message.
@madmax330 Should I use another message/method to check it?

**Comment by madmax330** [comment]
I think this message is fine since we already validate this type of field

**Comment by getusha** [comment]
`BUG:` Crash when sending invoice from chat 2nd time

1. Send invoice using the steps in the PR description
2. Press the `+` from the composer
3. Send Invoice > Enter Amount > Next > Next

Note: you may have to try again a couple of times

https://github.com/user-attachments/assets/cb2a0360-5bc9-4082-ad30-f2c86f1000dd

**Comment by getusha** [comment]
> We need the company name and URL to create a Stripe Connect account, and so we will ask this one time, on the very first invoice being sent from a workspace

@VickyStash It always navigates me to enter both fields as new, despite filling it for previous invoices.

https://github.com/user-attachments/assets/944ed177-648b-4279-9bfe-e1f499b30330

**Comment by VickyStash** [comment]
> @VickyStash It always navigates me to enter both fields as new, despite filling it for previous invoices.

It's cause you followed the **pre-step** described in the PR:

> !Pre-step: Update [hasInvoicingDetails check](https://github.com/Expensify/App/pull/46009/files#diff-c3278b76b0699dc5c04d47722c67669ef432ee715141b506716a6be212eb3f08R208-R212) to always return false so you can **always** see the expected screen. Please, remember that API doesn't support adding company name and the website for now.

@getusha So it always shows you the screen for testing purposes cause you updated it so. When the API updated the check will works as expected.

**Comment by VickyStash** [comment]
> `BUG:` Crash when sending invoice from chat 2nd time
> 
> 1. Send invoice using the steps in the PR description
> 2. Press the `+` from the composer
> 3. Send Invoice > Enter Amount > Next > Next
> 
> Note: you may have to try again a couple of times
> 
>  Screen.Recording.2024-08-06.at.2.25.10.in.the.afternoon.mov

@getusha It looks like this bug isn't related to this PR, see relevant issue https://github.com/Expensify/App/issues/46821

**Comment by getusha** [comment]
>It's cause you followed the pre-step described in the PR:

We will be handling that as a follow up? is it awaiting backend changes?

**Review by getusha — APPROVED** [review]

**Comment by VickyStash** [comment]
> We will be handling that as a follow up? is it awaiting backend changes?

Yes, I'll prepare a follow-up when API is ready

**Review by madmax330 — APPROVED** [review]

**Pull request merged**